alphadia/workflow/optimization.py

```diff
@@ -0,0 +1,684 @@
+                if self.estimator_group_name == "precursor"
+                else fragments_df
+            )
+
+            self.workflow.optimization_manager.fit({self.parameter_name: new_parameter})
+
+            self.reporter.log_string(
+                f"❌ {self.parameter_name:<15}: optimization incomplete after {len(self.history_df)} search(es). Will search with parameter {self.workflow.optimization_manager.__dict__[self.parameter_name]:.4f}.",
+                verbosity="progress",
+            )
+
+    def plot(self):
+        """Plot the value of the feature used to assess optimization progress against the parameter value, for each value tested."""
+        fig, ax = plt.subplots()
+
+        ax.axvline(
+            x=self.workflow.optimization_manager.__dict__[self.parameter_name],
+            ymin=0,
+            ymax=self.history_df[self.feature_name].max(),
+            color="red",
+            zorder=0,
+            label=f"Optimal {self.parameter_name}",
+        )
+
+        sns.lineplot(
+            data=self.history_df,
+            x="parameter",
+            y=self.feature_name,
+            ax=ax,
+        )
+        sns.scatterplot(
+            data=self.history_df,
+            x="parameter",
+            y=self.feature_name,
+            ax=ax,
+        )
+
+        ax.set_xlabel(self.parameter_name)
+        ax.xaxis.set_inverted(True)
+        ax.set_ylim(bottom=0, top=self.history_df[self.feature_name].max() * 1.1)
+        ax.legend(loc="upper left")
+
+        plt.show()
+
+    @abstractmethod
+    def _propose_new_parameter(self, df):
+        """This method specifies the rule according to which the search parameter is updated between rounds of optimization. The rule is specific to the parameter being optimized.
+
+        Parameters
+        ----------
+
+        df: pd.DataFrame
+            The dataframe used to update the parameter. This could be the precursor or fragment dataframe, depending on the search parameter being optimized.
+
+        Returns
+        -------
+        float
+            The proposed new value for the search parameter.
+
+
+        """
+        pass
+
+    @abstractmethod
+    def _check_convergence(self):
+        """This method checks if the optimization has converged according to parameter-specific conditions.
+
+        Returns
+        -------
+        bool
+
+        """
+        pass
+
+    @abstractmethod
+    def _get_feature_value(
+        self, precursors_df: pd.DataFrame, fragments_df: pd.DataFrame
+    ):
+        """Each parameter is optimized according to a particular feature. This method gets the value of that feature for a given round of optimization.
+
+        Parameters
+        ----------
+
+        precursors_df: pd.DataFrame
+            The precursor dataframe for the search
+
+        fragments_df: pd.DataFrame
+            The fragment dataframe for the search
+
+
+        """
+        pass
+
+
+class TargetedOptimizer(BaseOptimizer):
```

Comment:
`FixedToleranceOptimizer`? and rename `parameter` -> `tolerance`? (mb I am completely off)

Reply:
That could work, but I think it is OK as it is. Perhaps we can revisit this later.